block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala

```diff
@@ -0,0 +1,255 @@
+package coop.rchain.blockstorage
+
+import java.io.RandomAccessFile
+import java.nio.ByteBuffer
+import java.nio.file.{Files, Path}
+import java.util.stream.Collectors
+
+import cats.Monad
+import cats.effect.{Concurrent, ExitCase, Sync}
+import cats.implicits._
+import cats.effect.concurrent.{Ref, Semaphore}
+import com.google.protobuf.ByteString
+import coop.rchain.blockstorage.BlockStore.BlockHash
+import coop.rchain.blockstorage.FileLMDBIndexBlockStore.Checkpoint
+import coop.rchain.blockstorage.errors.{CheckpointsAreNotConsecutive, CheckpointsDoNotStartFromZero}
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.shared.Resources.withResource
+import coop.rchain.blockstorage.util.byteOps._
+import coop.rchain.shared.Log
+import coop.rchain.shared.ByteStringOps._
+import org.lmdbjava.DbiFlags.MDB_CREATE
+import org.lmdbjava._
+
+import scala.collection.JavaConverters._
+import scala.ref.WeakReference
+import scala.util.control.NonFatal
+import scala.util.matching.Regex
+
+class FileLMDBIndexBlockStore[F[_]: Monad: Sync] private (
+    lock: Semaphore[F],
+    env: Env[ByteBuffer],
+    index: Dbi[ByteBuffer],
+    blockMessageRandomAccessFileRef: Ref[F, RandomAccessFile],
+    checkpointsRef: Ref[F, List[Checkpoint]]
+) extends BlockStore[F] {
+  private[this] def withTxn[R](txnThunk: => Txn[ByteBuffer])(f: Txn[ByteBuffer] => R): F[R] =
+    Sync[F].bracketCase(Sync[F].delay(txnThunk)) { txn =>
+      Sync[F].delay {
+        val r = f(txn)
+        txn.commit()
+        r
+      }
+    } {
+      case (txn, ExitCase.Error(NonFatal(ex))) =>
+        Sync[F].delay {
+          ex.printStackTrace()
+          txn.close()
+        } *> Sync[F].raiseError(ex)
+      case (txn, _) => Sync[F].delay(txn.close())
+    }
+
+  private[this] def withWriteTxn(f: Txn[ByteBuffer] => Unit): F[Unit] =
+    withTxn(env.txnWrite())(f)
+
+  private[this] def withReadTxn[R](f: Txn[ByteBuffer] => R): F[R] =
+    withTxn(env.txnRead())(f)
+
+  private def readBlockMessage(offset: Long): F[BlockMessage] =
```

Comment:
every call to `RandomAccessFile` is just closed over `Sync[F]` which (as mentioned before) will happly rethorw exception (when `Task` is used as instance). 

why this is bad can simply be seen in usage of `readBlockMessage`. This private method is for example used in `get` method. If while calling `get` the `readBlockMessage` explosdes, so will the `get` method. Your `clear` function (that you've added in this PR) will never be executed.

Instread `readBlockMessage should return errors as values from which you could easily recover (compoiler will guide you). By closing it over `delay` you allow swallowing of every exception that might happen

Reply:
I have improved error handling in `readBlockMessage` (and other IO methods as well) so it returns `StorageIOErrT[F, BlockMessage]` now. I have achieved this through calling `attempt` on every `Sync[F].delay { ioStuff }` call and wrapping IOExceptions into a blockstore ADT of errors. Not sure if this is the best way, feel free to comment.

Also, I have left the ADT of errors a subclass of exception because some of the older code in `blockstorage` module still tries to throw it through `Sync[F].raiseError`. I won't address it in this PR to keep it on point but will get back to it in another PR.

Reply:
👍 
I'm more then happy to work with you on that with the next PR after this one